lms/djangoapps/completion/tests/test_models.py

```diff
@@ -142,3 +144,117 @@ def test_cannot_call_submit_completion(self):
                 completion=0.9,
             )
         self.assertEqual(models.BlockCompletion.objects.count(), 1)
+
+
+@ddt.ddt
+class AggregateCompletionTestCase(TestCase):
+
+    @classmethod
+    def setUpClass(cls):
+        super(AggregateCompletionTestCase, cls).setUpClass()
+        cls.overrider = waffle.waffle().override(waffle.ENABLE_COMPLETION_TRACKING, True)
+        cls.overrider.__enter__()
+
+    @classmethod
+    def tearDownClass(cls):
+        cls.overrider.__exit__(None, None, None)
+        super(AggregateCompletionTestCase, cls).tearDownClass()
+
+    def setUp(self):
+        super(AggregateCompletionTestCase, self).setUp()
+        self.user = UserFactory()
+
+    @ddt.data(
+        # Valid arguments
+        (u'block-v1:edx+test+run+type@video+block@doggos', 'course', 0.5, 1, 0.5),
+    )
+    @ddt.unpack
+    def test_submit_completion_with_valid_data(self, block_key, aggregate_name, earned, possible, expected_percent):
+        block_key_obj = UsageKey.from_string(block_key)
+        obj, is_new = models.AggregateCompletion.objects.submit_completion(
+            self.user,
+            block_key_obj.course_key,
+            block_key_obj,
+            aggregate_name,
+            earned,
+            possible
+        )
+        self.assertTrue(is_new)
+        self.assertEqual(len(models.AggregateCompletion.objects.all()), 1)
+        self.assertEqual(obj.earned, earned)
+        self.assertEqual(obj.possible, possible)
+        self.assertEqual(obj.percent, expected_percent)
+
+    @ddt.data(
+        # Earned greater than possible
+        (u'block-v1:edx+test+run+type@video+block@doggos', 'course', 1.1, 1),
+        # Earned is less than zero.
+        (u'block-v1:edx+test+run+type@video+block@doggos', 'course', -0.5, 1),
+        # Possible is less than zero.
+        (u'block-v1:edx+test+run+type@video+block@doggos', 'course', 0.5, -1),
+    )
+    @ddt.unpack
+    def test_submit_completion_with_exception(self, block_key, aggregate_name, earned, possible):
+        block_key_obj = UsageKey.from_string(block_key)
+        try:
+            models.AggregateCompletion.objects.submit_completion(self.user, block_key_obj.course_key,
+                                                                 block_key, aggregate_name, earned, possible)
+            self.fail()
+        except TypeError:
+            pass
+        except ValueError:
+            pass
+        except ValidationError:
+            pass
```

Comment:
This test is not testing what you think it's testing..

https://github.com/edx/edx-platform/pull/16715/commits/b31f3374967a0eab2c13a75c399e51128141182b#diff-41a6bf1a57b3cfe2e1e236483ddd3787 shows what exceptions are actually being raised, and reveals several issues, one of which serious:

* `TypeError` was actually being thrown for *all cases*, because you passed in `block_key` instead of `block_key_obj`.
* "Possible is less than zero." isn't tested, unless you also pass `earned < possible`, because it hits the `ValueError: The earned cannot be larger than the possible value` first.
* `ValidationError` is never thrown!  This is actually really important, and meant I had to look up stuff about [field validation in django models, which surprised the heck out of me](http://jamescooke.info/djangos-model-save-vs-full_clean.html), and made me kind of sad.
  Essentially, we have to call `full_clean` whenever we create one of these `AggregateCompletion` objects!

Fixed by using the `pre_save` hook: https://github.com/edx/edx-platform/pull/16715/commits/b31f3374967a0eab2c13a75c399e51128141182b#diff-e4200095aaff9a58d07a3a1c736d65ab.

Reply:
That's quite the deep dive. Good catches all around. That's interesting about the ValidationError issue.